pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -535,3 +539,96 @@ func (os *OpenStack) Routes() (cloudprovider.Routes, bool) {
 
 	return r, true
 }
+
+// Implementation of sort interface for blockstorage version probing
+type APIVersionsByID []apiversions_v1.APIVersion
+
+func (apiVersions APIVersionsByID) Len() int {
+	return len(apiVersions)
+}
+
+func (apiVersions APIVersionsByID) Swap(i, j int) {
+	apiVersions[i], apiVersions[j] = apiVersions[j], apiVersions[i]
+}
+
+func (apiVersions APIVersionsByID) Less(i, j int) bool {
+	return apiVersions[i].ID > apiVersions[j].ID
+}
+
+func (os *OpenStack) volumeService() (volumeService, error) {
+	switch os.bsOpts.BSVersion {
+	case "v1":
+		sClient, err := openstack.NewBlockStorageV1(os.provider, gophercloud.EndpointOpts{
+			Region: os.region,
+		})
+		if err != nil || sClient == nil {
+			glog.Errorf("Unable to initialize cinder client for region: %s", os.region)
+			return nil, err
+		}
+		return &VolumesV1{sClient, os.bsOpts}, nil
+	case "v2":
+		sClient, err := openstack.NewBlockStorageV2(os.provider, gophercloud.EndpointOpts{
+			Region: os.region,
+		})
+		if err != nil || sClient == nil {
+			glog.Errorf("Unable to initialize cinder v2 client for region: %s", os.region)
+			return nil, err
+		}
+		return &VolumesV2{sClient, os.bsOpts}, nil
+	case "auto":
+		sClient, err := openstack.NewBlockStorageV1(os.provider, gophercloud.EndpointOpts{
+			Region: os.region,
+		})
+		if err != nil || sClient == nil {
+			glog.Errorf("Unable to initialize cinder client for region: %s", os.region)
+			return nil, err
+		}
+
+		err = apiversions_v1.List(sClient).EachPage(func(page pagination.Page) (bool, error) {
+			apiversions, err := apiversions_v1.ExtractAPIVersions(page)
+			if err != nil {
+				glog.Errorf("Unable to list blockstorage API versions for region: %s", os.region)
+				return false, err
+			}
+
+			sort.Sort(APIVersionsByID(apiversions))
+			// take first CURRENT or SUPPORTED version
+			for _, version := range apiversions {
+				if version.Status == "CURRENT" || version.Status == "SUPPORTED" {
```

Comment:
I still think it should only be checking "CURRENT" here (thereby removing the utility of sorting the versions). To me, for anything other than testing, users should explicitly specify the Block Storage (or, indeed, any other service) version they want; if they don't, they should get the current version or an error.

Alternatively, you could check for current versions first, then fall back to supported versions (untested):
```go
for _, status := range []string{"CURRENT", "SUPPORTED"} {
  for _, version := range apiversions {
    if version.Status == status {
	switch version.ID {
	case "v2.0":
		os.bsOpts.BSVersion = "v2"
		return false, nil
	case "v1.0":
		os.bsOpts.BSVersion = "v1"
		return false, nil
	}
    }
  }
}
```